### PR TITLE
Add post contract execution logic to handle gas payment

### PR DESCRIPTION
**Pull request by cowboy-bebug**
Implement `GasHandler` trait for Runtime as the post contract execution. It tries to access storage with `cennzx-spot` as key to get FeeExchange, which is consumed at `buy_fee_exchange`.

### CODE REVIEW

File: runtime/src/lib.rs [hunk 1]
```diff
@@ -585,6 +589,32 @@ impl additional_traits::DelegatedDispatchVerifier for Runtime {
 	}
 }
 
+/// Handles gas payment post contract execution (before deferring runtime calls) via CENNZX-Spot exchange.
+impl<T> pallet_contracts::GasHandler<T> for Runtime
+where
+	T: pallet_contracts::Trait + pallet_generic_asset::Trait + crml_cennzx_spot::Trait + Decode,
+{
+	fn empty_unused_gas(transactor: &T::AccountId, gas_meter: GasMeter<T>) {
+		let key = &b"cennzx-spot"[..];
```
Comment (jordy25519) [inline runtime/src/lib.rs]: Could throw this into constants.rs and would prefer a name related to fee exchange / gas

File: runtime/src/lib.rs [hunk 2]
```diff
@@ -585,6 +589,32 @@ impl additional_traits::DelegatedDispatchVerifier for Runtime {
...
+
+		// The take() function ensures the entry is "killed" after access.
+		// Note: if there's no entry (ie, None), it means we handled an edge case of
+		// CPAY in FeeExchange, ie, gas is already paid in CPAY.
+		if let Some(exchange_op) = unhashed::take::<FeeExchange>(&key) {
+			let gas_spent = gas_meter.spent();
+
+			// Fee exchange can never fail as conditions such as having enough liquidity
+			// are checked early (before FeeExchange is put into database)
```
Comment (jordy25519) [inline runtime/src/lib.rs]: for consistency `database` -> `storage`

File: runtime/src/lib.rs [hunk 3]
```diff
@@ -585,6 +589,32 @@ impl additional_traits::DelegatedDispatchVerifier for Runtime {
...
 
+/// Handles gas payment post contract execution (before deferring runtime calls) via CENNZX-Spot exchange.
+impl<T> pallet_contracts::GasHandler<T> for Runtime
+where
+	T: pallet_contracts::Trait + pallet_generic_asset::Trait + crml_cennzx_spot::Trait + Decode,
+{
+	fn empty_unused_gas(transactor: &T::AccountId, gas_meter: GasMeter<T>) {
+		let key = &b"cennzx-spot"[..];
+
+		// The take() function ensures the entry is "killed" after access.
```
Comment (jordy25519) [inline runtime/src/lib.rs]: Don't forget the back ticks ` around keywords

File: runtime/src/lib.rs (hunk 1 above)
Comment (cowboy-bebug) [reply runtime/src/lib.rs]: We will use something like this: `const GAS_FEE_EXCHANGE = "cennzx-spot"`

Comment (jordy25519) [reply runtime/src/lib.rs]: can you change the actual key byte string to match the variable name also?

File: runtime/src/lib.rs [hunk 4]
```diff
@@ -585,6 +589,30 @@ impl additional_traits::DelegatedDispatchVerifier for Runtime {
 	}
 }
 
+/// Handles gas payment post contract execution (before deferring runtime calls) via CENNZX-Spot exchange.
+impl<T> pallet_contracts::GasHandler<T> for Runtime
+where
+	T: pallet_contracts::Trait + pallet_generic_asset::Trait + crml_cennzx_spot::Trait + Decode,
+{
+	fn empty_unused_gas(transactor: &T::AccountId, gas_meter: GasMeter<T>) {
```
Comment (aliXsed) [inline runtime/src/lib.rs]: How have you made sure that this function would be called for a contract?

Comment (cowboy-bebug) [reply runtime/src/lib.rs]: Thanks for spotting this!

File: crml/cennzx-spot/src/lib.rs [hunk 5]
```diff
@@ -729,7 +729,7 @@ impl<T: Trait> Module<T> {
 	}
 
 	/// `asset_id` - Trade asset
-	/// `amount_sold` - Amount of trade assets sold
+	/// `amount_sold` - Amount of the trade asset to sell
```
Comment (aliXsed) [inline crml/cennzx-spot/src/lib.rs:732]: @hoani @royyang-centrality @holygits Do you think I should revert this change? I now think I didn't get what that sentence meant (still not getting it to be honest)

Comment (roy-powerfinance) [reply crml/cennzx-spot/src/lib.rs:732]: I think your comment is correct. Here we use Trade asset to buy Core, and we specify the amount of trade asset we are willing to use, and expect the function to return the amount of Core we can buy.
I.e. I have 100 sylo, tell me how much Cpay I can get out of this

Comment (roy-powerfinance) [reply crml/cennzx-spot/src/lib.rs:732]: the parameter name in quote should say `sell_amount`

**Comment by jordy25519** [comment]
Superseded by #133
